wallet-new/server/Cardano/Wallet/API/V1/LegacyHandlers/Addresses.hs

```diff
@@ -25,21 +30,32 @@ handlers =  listAddresses
        :<|> newAddress
        :<|> verifyAddress
 
+-- | This is quite slow. What happens when we have 50k addresses?
+-- TODO(ks): One idea I have is to persist the length of the
+-- addresses and send that to `respondWith`,
+-- while we could lazily fetch the data (and have better performance)
+-- we need to show what the current page contains?
+-- In other words, we wouldn't send everything to `respondWith`.
+-- Another idea is to use actual paging to reduce the footprint of all
+-- these calls. The flaw with this is that we need to think about deletion,
+-- but I have an idea or two how that can be fixed.
 listAddresses
-    :: MonadIO m
-    => RequestParams -> m (WalletResponse [Address])
-listAddresses RequestParams {..} = do
-    addresses <- liftIO $ generate (vectorOf 2 arbitrary)
-    return WalletResponse {
-              wrData = addresses
-            , wrStatus = SuccessStatus
-            , wrMeta = Metadata $ PaginationMetadata {
-                        metaTotalPages = 1
-                      , metaPage = 1
-                      , metaPerPage = 20
-                      , metaTotalEntries = 2
-                      }
-            }
+    :: (MonadThrow m, V0.MonadWalletLogic ctx m)
+    => RequestParams
+    -> m (WalletResponse [WalletAddress])
+listAddresses params = do
+
+    wdb <- askWalletDB
+    ws  <- getWalletSnapshot wdb
+
+    let walletAddresses   = getWalletAddresses ws
+    allCAccounts         <- concatMapM (V0.getAccounts . Just) walletAddresses
+    let allCAddresses     = concatMap caAddresses allCAccounts
+    let allAddresses      = migrate allCAddresses
```

Comment:
Why not doing everything in one "gulp" instead of three separate passes over the input data?

Reply:
What you lose though is efficiency -- we are doing 3 linear scans for a potentially-huge list of addresses 😉

Reply:
List fusion should work out here, if migrate is a good consumer.

Reply:
@parsonsmatt Good idea, thanks.

Reply:
My point being: relying on list fusion to trigger is dangerous and we should not assume it will 😉 But I don't think at this stage that's worthwhile getting into the premature optimisation trap -- let's see how it goes, and let's optimise later.